### DISCUSSION TITLE
Modulation state immediately after note on

### DISCUSSION THREAD

**Discussion by swesterfeld** [free-audio/clap#316] [Q&A]
I am currently working on CLAP support for my SpectMorph plugin. I'd like to support per-note modulation as well. One issue I don't know how to deal with is that I've seen that Bitwig only sends modulation state updates every 64 samples. So this means that usually at the point where the note on event happens (these are sample accurate), I have no modulation state.

So for instance I get 
- a note on at time 1000 with a clap_id=42
- a modulation update for some parameter at time 1024 for clap_id=42

So this means that for that parameter, there is a small amount of samples (24 in this example) that I have to process without knowing the modulation for this parameter. So the parameter will be rendered without any modulation for 24 samples, an then the modulation will make the parameter jump to some state due to modulation and then I get periodic modulation updates.

Now one could say that the time before the first modulation event arrives should be mostly inaudible, because 24 samples are really short. However, internally SpectMorph uses additive synthesis via inverse FFT, so to render the first 24 samples, it will fill up a whole buffer of audio data internally. The parameters to build that buffer are computed without any modulation (as for the very first sample, there is no modulation available). This means that there is a whole FFT buffer that is computed without any modulation, and this can be audible.

I am not sure how to proceed here, as I see no way of avoiding these audio glitches by doing something clever within my plugin? One solution would be to specify in the CLAP spec that hosts should always provide the modulation state of all parameters immediately after note on?

**Comment by robbert-vdh** [comment]
See https://github.com/free-audio/interop-tracker/issues/51.

**Reply by swesterfeld** [reply]
Yes, this looks exactly like my problem.

**Reply by robbert-vdh** [reply]
This could be worked around by keeping track of the last polyphonic modulation events for voices that do not (yet) exist and then looking up the current values when starting a voice (making sure to clear those events again so they won't be reused if the host somehow reuses voice IDs), but this is definitely a host issue that should be fixed.

**Reply by swesterfeld** [reply]
No this strategy would not work. For instance if you have an "Expressions" modulator and use the velocity of the event as polyphonic modulation for a parameter, the last polyphonic modulation event before the note on will have a value of zero. So the first valid modulation event only occurs after the note on.
